### PR TITLE
Feature/tech 701 parallel improvements

### PR DESCRIPTION
**Pull request by Jorg88**
----
### 📕 [TECH-701](https://vandebron.atlassian.net/browse/TECH-701) Look into parallel stage steps jorgpost@vandebron.nl 
Investigate if parallel runs are an option for the build and test stages. Maybe even configurable (like name & icon) from the run properties, i.e. set the amount of threads there.

There is example code in `src/mpyl/steps/postdeploy/cypress_test.py` which can maybe we extracted and re-used, keep in mind that code is for docker cmd’s though. Probably needs to be made a bit more abstract.

If this works, see how this affects build times. Start with only 2 threads and ramp up until performance gets worse because of resource limits in Jenkins / Platform. 

🏗️ Build [8](https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline%20-%20Test/job/PR-337/8/display/redirect) ✅ Successful, started by _Jorg Post_  
🚀 *[cloudfront-service](https://cloudfront-service-337.test.nl/swagger/index.html)*, *example-dagster-user-code*, *job*, *kong-sync*, *[nodeservice](https://nodeservice-337.test.nl/swagger/index.html)*, *ocpp-first*, *ocpp-second*, *[sbtservice](https://sbtservice-337.test.nl/swagger/index.html)*, *sparkJob-first*, *sparkJob-second*  

[TECH-701]: https://vandebron.atlassian.net/browse/TECH-701?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: releases/notes/1.4.12.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#### Enhancements
+- Several small code improvements
+
+#### Bug Fixes
+- Fix `mpyl build clean` command for override projects
```
Comment (koeves) [inline releases/notes/1.4.12.md:5]: And also make mpyl parallel ready?

Comment (koeves) [reply releases/notes/1.4.12.md:5]: or you took those out completely I see

File: src/mpyl/utilities/parallel/__init__.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    executor = ThreadPoolExecutor(max_workers=number_of_threads)
+    for command in commands:
+        threads.append(executor.submit(command.function, **command.parameters))
+
+    results: list[T] = list(__flatten([thread.result() for thread in threads]))
+
+    return results
+
+
+def __flatten(object_to_flatten: Any):
```
Comment (koeves) [inline src/mpyl/utilities/parallel/__init__.py:29]: Is this more readable than the itertools.chain?

**Review by koeves — APPROVED** [review]

File: src/mpyl/utilities/parallel/__init__.py (hunk 2 above)
Comment (Jorg88) [reply src/mpyl/utilities/parallel/__init__.py:29]: I don't remember the details, but there was a reason I used this. Not sure if `itertools.chain` didn't do precisely what we need or Sam commented something

Comment (koeves) [reply src/mpyl/utilities/parallel/__init__.py:29]: could be that the object is not always an iterable? I would then move the if to the outside and call itertools.chain

File: src/mpyl/steps/test/sbt.py [hunk 3]
```diff
@@ -93,7 +94,10 @@ def execute(self, step_input: Input) -> Output:
             suite = to_test_suites(
                 cast(JunitTestSpec, test_result.produced_artifact.spec)
             )
-            summary = sum_suites(suite)
+            errors = sum_suites(suite).errors
+            summary = dataclasses.replace(
+                sum_suites(suite), errors=errors if test_result.success else errors + 1
+            )
```
Comment (koeves) [inline src/mpyl/steps/test/sbt.py:106]: I would refactor this so we only call sum_suites once

Comment (koeves) [reply src/mpyl/steps/test/sbt.py:106]:
```suggestion
            combined = sum_suites(suite)
            errors = combined.errors
            summary = dataclasses.replace(
                combined, errors=errors if test_result.success else errors + 1
            )
```

File: src/mpyl/utilities/subprocess/__init__.py [hunk 4]
```diff
@@ -9,7 +9,10 @@
 
 
 def custom_check_output(
-    logger: Logger, command: Union[str, list[str]], capture_stdout: bool = False
+    logger: Logger,
+    command: Union[str, list[str]],
+    capture_stdout: bool = False,
+    use_print: bool = False,
```
Comment (koeves) [inline src/mpyl/utilities/subprocess/__init__.py:15]: Maybe `stream_output` instead of print? but both are fine

Comment (Jorg88) [reply src/mpyl/utilities/subprocess/__init__.py:15]: but both print and log can be / are streams?

Comment (koeves) [reply src/mpyl/utilities/subprocess/__init__.py:15]: fair enough

File: src/mpyl/steps/test/sbt.py [hunk 5]
```diff
@@ -102,23 +91,22 @@ def execute(self, step_input: Input) -> Output:
...
-        step_input: Input, config: SbtConfig, commands_fn: Callable[[Input], list[str]]
+        project_name: str, config: SbtConfig, compile_test: bool
     ):
-        return config.to_command(config.test_with_client, commands_fn(step_input))
+        command = list(
+            filter(
+                None,
+                [
+                    f"project {project_name}",
+                    "coverageOn" if config.test_with_coverage else "",
```
Comment (koeves) [inline src/mpyl/steps/test/sbt.py:103]: Use None instead of ""

**Review by koeves — APPROVED** [review]

**Pull request merged**